drivers/include/dcf77.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_dcf77 DCF77 long wave receiver with 77,5 kHz
+ * @ingroup     drivers_sensors
+ * @brief       Device driver long wave receiver with 77,5 kHz
+ *
+ * @{
+ *
+ * @file
+ * @brief       Interface definition for the dcf77 sensor driver
+ *
+ * @author      Michel Gerlach <michel.gerlach@haw-hamburg.de>
+ */
+
+#ifndef DCF77_H
+#define DCF77_H
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include "xtimer.h"
+#include "time.h"
+#include "periph/gpio.h"
+#include "dcf77_internal.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* forward-declaration for dcf77_tick_cb_t */
+struct dcf77;
+
+/**
+ * @brief Signature for tick callback
+ *
+ * @param[in] dev           device that triggered the alarm
+ * @param[in] arg           optional argument to put the callback in the right context
+ */
+typedef void(*dcf77_tick_cb_t)(struct dcf77 *dev, void *arg);
+
+/**
+ * @brief   Possible return codes
+ */
+enum {
+    DCF77_OK          =  0,       /**< all good */
+    DCF77_NOCSUM      = -1,       /**< checksum error */
+    DCF77_TIMEOUT     = -2,       /**< communication timed out */
+    DCF77_INIT_ERROR  = -3        /**< Initialization error */
+};
+
+/**
+ * @brief   Configuration parameters for DCF77 devices
+ */
+typedef struct {
+    gpio_t pin;             /**< GPIO pin of the device's data pin */
+    gpio_mode_t in_mode;    /**< input pin configuration from the device,
+                             * without pull resistor */
+} dcf77_params_t;
+
+
+/**
+ * @brief   Device descriptor for DCF77 sensor devices
+ */
+typedef struct dcf77 {
+    dcf77_params_t params;      /**< Device parameters */
+    dcf77_bits_t bitseq;        /**< contains all Bits from a current cycle */
+    dcf77_bits_t last_bitseq;   /**< contains all Bits from a last cycle */
+    uint8_t internal_state;     /**< internal States  */
+    uint32_t startTime;         /**< Timestamp to measure the term of the level */
```

Comment:
One last thing: If you swap the order of these, the struct will use [less memory](http://www.catb.org/esr/structure-packing/).

Reply:
You can squash directly :)

Reply:
exchanged positions

Reply:
Thanks for the many hints. As Riot-newby, that helps me enormously.

Reply:
Then I have another one: [Compiler Explorer](https://godbolt.org/) is a good tool to check how much memory a struct uses and more! If yo want to just check the size, write a function that just returns `sizeof(struct ...)`, but of course you can also check how the compiler will optimize your functions or how the generated code differs on different architectures.